### PR TITLE
Port Order to Dotty

### PR DESCRIPTION
**Pull request by andrzejressel**

### CODE REVIEW

**Comment by andrzejressel** [comment]
After merge I'll propably make every derivation lazy

**Comment by joroKr21** [comment]
> After merge I'll propably make every derivation lazy

WDYM? 🤔

**Comment by andrzejressel** [comment]
I basically mean replacing:
- `using inst: K0.ProductInstances[Or, A]` with `using inst: => K0.ProductInstances[Or, A]`
- `using using inst: K0.CoproductInstances[Or, A]` with `using inst: => K0.CoproductInstances[Or, A]`

And the same for `K1` and `Instances[Or, A]`.

**Comment by joroKr21** [comment]
> So recursive data classes will work properly.

I think @TimWSpence tried that in #401 but for some reason it's still not working properly

**Comment by andrzejressel** [comment]
Yeah, i saw that. I'll check if scala/shapeless update will fix that.

File: core/src/main/scala-3/cats/derived/DerivedOrder.scala [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+type DerivedOrder[A] = Derived[Order[A]]
+object DerivedOrder:
+  type Or[A] = Derived.Or[Order[A]]
+
+  inline def apply[A]: Order[A] =
+    import DerivedOrder.given
+    summonInline[DerivedOrder[A]].instance
+
+  @targetName("product")
+  given [A](using inst: => K0.ProductInstances[Or, A]): DerivedOrder[A] =
```
Comment (joroKr21) [inline core/src/main/scala-3/cats/derived/DerivedOrder.scala:20]: Let's just use the name
```suggestion
  given product[A](using inst: => K0.ProductInstances[Or, A]): DerivedOrder[A] =
```

File: core/src/main/scala-3/cats/derived/DerivedOrder.scala [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    given K0.ProductInstances[Order, A] = inst.unify
+    new Product[Order, A] {}
+
+  @targetName("coproduct")
+  given [A](using inst: => K0.CoproductInstances[Or, A]): DerivedOrder[A] =
```
Comment (joroKr21) [inline core/src/main/scala-3/cats/derived/DerivedOrder.scala:25]:
```suggestion
  given coproduct[A](using inst: => K0.CoproductInstances[Or, A]): DerivedOrder[A] =
```

File: core/src/test/scala-3/cats/derived/OrderSuite.scala [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+import cats.Order
+import cats.kernel.laws.discipline.{OrderTests, SerializableTests}
+import org.scalacheck.Arbitrary
+
+class OrderSuite extends KittensSuite {
+  import OrderSuite._
+  import TestDefns._
+
+  def testOrder(context: String)(using
```
Comment (joroKr21) [inline core/src/test/scala-3/cats/derived/OrderSuite.scala:27]: Same here - we can simplify by making this `inline` like with your `Apply` PR

**Review by joroKr21 — COMMENTED** [review]
Thanks again - this is amazing! ❤️

File: core/src/test/scala-3/cats/derived/OrderSuite.scala (hunk 3 above)
Comment (andrzejressel) [reply core/src/test/scala-3/cats/derived/OrderSuite.scala]: Here we don't need any Isomorphism and inline trick works great 👍

File: core/src/main/scala-3/cats/derived/DerivedOrder.scala (hunk 2 above)
Comment (andrzejressel) [reply core/src/main/scala-3/cats/derived/DerivedOrder.scala]: Fixed

**Review by joroKr21 — APPROVED** [review]
Thank you! One step closed to feature parity 👍

**Pull request merged**